session.go

```diff
@@ -0,0 +1,202 @@
+// This is needed so functions that enable sending relays can be provided with a struct that contains both session data and the endpoint(s) selected for receiving relays.
+type FilteredSession struct {
+	*sessiontypes.Session
+
+	// selectedEndpoints is the set of a specific service's endpoints, from the underlying session, that are selected by the user of FilteredNodes.
+	// This selection can be for any purpose, e.g. sending relays.
+	// The set of selected endpoints is stored as a map of supplier addresses to a list of endpoints.
+	selectedEndpoints map[string][]*sharedtypes.SupplierEndpoint
+
+	// selectedServiceId stores the ID of the service selected by the user.
+	selectedServiceId string
+}
+
+// TODO_DISCUSS: Consider using a custom type, defined as a string, as supplier address.
+// This can help enforce type safety by requiring explict type casting of a string before it can be used as a SupplierAddress.
+//
+// ServiceEndpoints returns the supplier endpoints assigned to a session for the given service id.
+//
+// The returned value is a map of supplierId to the list of endpoints.
+func (f *FilteredSession) ServiceEndpoints(
+	serviceId string,
+) (map[string][]*sharedtypes.SupplierEndpoint, error) {
+	if f.Session == nil {
+		return nil, fmt.Errorf("ServiceEndpoints: Session not set on FilteredSession struct")
+	}
+
+	supplierEndpoints := make(map[string][]*sharedtypes.SupplierEndpoint)
+	for _, supplier := range f.Session.Suppliers {
+		// TODO_DISCUSS: It may be a good idea to use a map[serviceID][]Endpoint under each supplier,
+		//   especially if service IDs under each supplier should be unique.
+		//   This would enable easy lookups based on service ID, without impacting any use cases that
+		//   need to iterate through all the services.
+		//
+		// The endpoins slice is intentionally defined here to prevent any overwrites in the unlikely case
+		//   that there are duplicate service IDs under a supplier.
+		var endpoints []*sharedtypes.SupplierEndpoint
+		for _, service := range supplier.Services {
+			// TODO_DISCUSS: considering a service Id is required to get a session, does it make sense for the suppliers under the session
+			// to only contain endpoints for the (single) service covered by the session?
+			if service.Service.Id != serviceId {
+				continue
+			}
+
+			endpoints = append(endpoints, service.Endpoints...)
+		}
+		supplierEndpoints[supplier.Address] = endpoints
+	}
+
+	return supplierEndpoints, nil
+}
+
+// SelectEndpoint adds the specifed supplier+endpoint combination to the list of selected endpoints of a service, e.g. for sending relays.
+// This method is not safe to use concurrently by multiple goroutines.
+func (f *FilteredSession) AddEndpointToSelection(serviceId string, supplierAddress string, endpoint *sharedtypes.SupplierEndpoint) error {
```

Comment:
I see a few problems with this method.
1. It adds unneeded statefulness to the `FilteredSession`. I see that passing a filter function would be more elegant.
2. It can accumulate selected `SupplierEndpoint`s but I see no method to retrieve them all, `SelectedEndpoint()` only returns the first.
3. It implies that the caller has already selected the endpoint to add and has the responsibility to pass in the matching `supplierAddress` (I really prefer a custom struct that contains wraps the endpoint with its corresponding supplier if we do not opt of passing a filter function).